warn user when sending from different account

brad-decker opened the pull request:
task: Warn the user when the "from" account of confirmation is different from the selected account

Trello card: https://trello.com/c/ZijoEYIz
Figma: https://www.figma.com/file/CvD7nhQVeSEB1QuSy3XhtY/oCap-%2F-Login-Per-Site?node-id=2783%3A144

<details>
<summary>Screenshot</summary>
<img src="https://user-images.githubusercontent.com/4448075/82078913-93b46780-96a7-11ea-82a7-59d606e63d82.png" />
</details>

brad-decker commented on ui/app/components/ui/sender-to-recipient/sender-to-recipient.component.js:
```diff
@@ -1,211 +1,265 @@
+  const permittedAccountsByOrigin = useSelector(getPermittedAccountsByOrigin)
+  const selectedAddress = useSelector(getSelectedAddress)
+
+  console.log(originOfCurrentTab)
```
:eyes:

brad-decker commented on ui/app/components/ui/sender-to-recipient/sender-to-recipient.component.js:
```diff
@@ -1,211 +1,265 @@
+  let showWarning = false
+  console.log(permittedAccounts, senderAddress)
+  if (permittedAccounts?.length > 0 && permittedAccounts.includes(senderAddress) && !permittedAccounts.includes(selectedAddress)) {
+    console.log('yo')
```
:eyes: 👋

brad-decker commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Warn the user when the "from" account of a confirmation is different from the selected account](https://trello.com/c/ZijoEYIz/70-warn-the-user-when-the-from-account-of-a-confirmation-is-different-from-the-selected-account)

Gudahtt commented on ui/app/components/ui/sender-to-recipient/index.scss:
```diff
@@ -17,6 +17,10 @@
 
       &__tooltip-container {
         max-width: 100%;
+        &--icon {
```
Nit: This doesn't seem very BEM like - `icon` doesn't seem like a modifier.

Maybe this is better?
```suggestion
        &-icon {
```

brad-decker replied:
replicated this pretty heavily. i'll resolve it throughout.

Gudahtt reviewed (changes requested):
Looks good aside from the missing `checksummedSenderAddress ` prop!

brad-decker commented on app/_locales/en/messages.json:
```diff
@@ -1033,6 +1033,9 @@
     "message": "No Thanks"
   },
+  "notCurrentAccount": {
+    "message": "Make sure this is the account you meant to send from! It's different than the currently selected account in your wallet."
```
@Gudahtt i used the messaging you suggested on the figma, but with the conversation, we just had about origin included in this maybe we should do that? cc @rachelcope

Gudahtt replied:
The `origin` should be featured heavily in all of our confirmation screens already. I think we're covered here already.

brad-decker commented on ui/app/hooks/useDappAuthorizationCheck.js:
```diff
@@ -0,0 +1,24 @@
+  const txData = useSelector(txDataSelector)
+
+  // Should this fallback to originOfCurrentTab? Not sure I understand this
+  const origin = txData?.origin || txData?.msgParams?.origin || originOfCurrentTab || undefined
```
@Gudahtt i could use an extra set of eyes on this hook.

first, some confirmations have the origin at the top level of txData, some have it in msgParams, and I'm unsure if I should be using originOfCurrentTab here, it seems useful in the event this hook would be used outside of a confirmation screen (if I understand getOriginofCurrentTab's intent correctly)

Secondly, I'm unsure of the name, return values, etc for this hook. It's nice to have all of this logic bundled together for use in the various confirmation screens but I want to make sure it makes sense. Also if we do decide to throw the origin into the tooltip message, we might want to return the actual warning message from this hook? if not that, then return the origin itself?

Gudahtt replied:
The "currentTab"-related stuff should probably be renamed to "activeTab" - that all refers to [this extension API feature](https://developer.chrome.com/extensions/activeTab), which lets you see the active tab when the browser action popup is triggered (e.g. when you click the MetaMask icon in the toolbar).

This is only set in the popup UI, so we only use it for popup-specific UI (like the connected status indicator). It shouldn't be used for confirmation screens at all, since those are typically in the notification popup, where there is no active tab.

I'll have to look a bit further into how this hook is being used, but I don't think there are any scenarios where we'd want to fallback from some other origin to the active tab origin. All transactions we receive from a dapp should have an origin associated with them - otherwise something is broken. I thought that even the send flow transactions had _something_ set as the origin :thinking:

Gudahtt replied:
I'm a bit confused by the `dappIsAuthorizedForAccount` check - the dapp should _always_ be authorized for this account, otherwise it wouldn't be possible for us to get here. Is there a case you found where a dapp was able to suggest a transaction for an account they aren't permitted to interact with?

Gudahtt replied:
Oh, I guess I just referenced one case: the send flow. Since any transactions created via the send flow are directly proposed by the user, not the dapp, it doesn't make sense to verify the origin is permitted for those. The interaction in that case is entirely independent of the dapp.

brad-decker replied:
hmm, I think I might have confused myself when trying to name this API after having written the logic. The check against the domain is really there to see if the request is being initiated by a Dapp the user has connected to in any of their accounts. Maybe I went to overboard with logic here and it's really as simple as 

```
const showWarning = selectedAccount !== fromAccount;
```

brad-decker replied:
bc, the only case where the fromAccount would be different from selected is in the event of a dapp initiated transaction

Gudahtt replied:
I think it may be as simple as `const showWarning = selectedAccount !== fromAccount;` - even the send flow example I thought of earlier doesn't really matter - we can still show this warning if the "from" account differs from the selected account in the send flow, because that should rarely happen anyway.

brad-decker commented on ui/app/components/ui/sender-to-recipient/sender-to-recipient.component.js:
```diff
@@ -1,211 +1,262 @@
+        senderAddressCopied
+          ? <p>{t('copiedExclamation')}</p>
+          : addressOnly
+            ? <p>{t('copyAddress')}</p>
```
I mean, why not fix it up while I'm there.

Gudahtt commented on ui/app/components/ui/sender-to-recipient/sender-to-recipient.component.js:
```diff
@@ -1,211 +1,265 @@
-        </div>
+        )}
+        <SenderAddress
+          checksummedRecipientAddress={checksummedRecipientAddress}
```
I think this was supposed to be `checksummedSenderAddress` instead of `checksummedRecipientAddress`

```suggestion
          checksummedSenderAddress={checksummedSenderAddress}
```

brad-decker committed:
refactor to functional components, drop proxyquire

brad-decker committed:
add new info icon component

brad-decker committed:
Add new account mismatch warning component

brad-decker committed:
include scss from new components

brad-decker committed:
add new warning component to conf screens

brad-decker committed:
fix test

brad-decker commented:
@Gudahtt could use a re-review when you have time, included a commit breakdown in the description.

rekmarks reviewed (changes requested):
Looks great, and I really like the refactor!

However, the address "copied to clipboard" tooltip doesn't appear to be working correctly. Notice in the recording how the sender tooltip changes without clicking, then both tooltips display `Copied!` regardless of what I click.

<details>
<summary> Screen Recording </summary>
<img src="https://user-images.githubusercontent.com/25517051/82498963-acf15580-9aa5-11ea-9b07-2681a0b25a69.gif"/>
</details>

rachelcope commented:
What about updating the copy to something a little more concise?

"Is this the correct account? It's different from the currently selected account in your wallet"

@brad-decker @Gudahtt

brad-decker commented:
@rekmarks -- Thanks for finding that! The current experience on develop is actually pretty broken. First off, the hover state only works half of the time for the recipient address hover... only if you go from sender tooltip active -> move the mouse over directly to the recipient. If you enter the recipient address from anywhere else it doesn't work. Furthermore, clicking it doesn't update the message to say "Copied!" because it doesn't actually trigger a set state call. 

To address this:
1. updated react-tippy
2. added an offset prop to tooltip-v2 (because the reason for flakey hover seems to be related to the tooltip trying to open partially offscreen?)
3. separate state for recipient and sender address copying
4. Move more of the logic around in the split up components so that state doesn't need to be passed anywhere. 

all of this is isolated to the latest commit.

brad-decker committed:
change message copy

brad-decker commented:
@rachelcope made the copy change.

brad-decker committed:
removing new instances of useContext(I18nContext)

brad-decker commented on package.json:
```diff
@@ -160,7 +160,7 @@
     "react-select": "^1.0.0",
     "react-simple-file-input": "^2.0.0",
-    "react-tippy": "^1.2.2",
+    "react-tippy": "1.4.0",
```
@Gudahtt I did another test against 1.2.2 to make sure I absolutely required the upgrade. I don't believe I do. Used prod build to determine if it worked at 1.2.2 and it does. sometimes developer build would periodically not work and I am not sure why... regardless the changes to deps is removed

rekmarks reviewed (approved):
LGTM!

Gudahtt reviewed (approved):
LGTM!
